File: app/src/main/java/util/ContentReader.java
package util;

import java.io.BufferedReader;
import java.io.Closeable;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;
import java.util.stream.Stream;

import dictionary.Word;

public class ContentReader implements Closeable {
    private final BufferedReader reader;

    public ContentReader(InputStream content) {
        reader = new BufferedReader(new InputStreamReader(content, StandardCharsets.UTF_8));
    }

    public List<Word> readContent() throws IOException {
        List<Word> content = new ArrayList<>();
        while (reader.ready()) {
            String line = reader.readLine();
            if (!line.isBlank()) {
                String[] tokens = line.split("::");
                if (tokens.length == 2) {
                    String[] words = tokens[0].split(", ");
                    List<String> inflections = Stream.of(words).skip(1).collect(Collectors.toList());
                    content.add(new Word(words[0], tokens[1], inflections));
                } else {
                    throw new IOException("Invalid input: " + line);
                }
            }
        }

        return content;
    }

    @Override
    public void close() throws IOException {
        reader.close();
    }
}

File: app/src/main/java/util/ContentReader.kt
package util

import dictionary.Word
import java.io.*

/**
 * Reader to read the Words with their inflections and definitions from a file
 */
class ContentReader(content: InputStream, private val delimiter: String = "::", private val inflectionDelimiter: String = ",") : Closeable {
    private val reader: BufferedReader

    init {
        reader = content.bufferedReader()
    }

    @Throws(IOException::class)
    fun readContent(): List<Word> {
        val content: MutableList<Word> = ArrayList()
        while (reader.ready()) {
            val line = reader.readLine()
            if (line.isNotBlank()) {
                val tokens = line.split(delimiter)
                if (tokens.size == 2) {
                    val words = if(tokens[0].startsWith("“")) listOf(tokens[0]) else tokens[0].split(inflectionDelimiter)
                    val inflections = words.asSequence().drop(1).toList()
                    content.add(Word(words[0], tokens[1], inflections))
                } else {
                    throw IOException("Invalid input: $line")
                }
            }
        }
        return content
    }

    @Throws(IOException::class)
    override fun close() {
        reader.close()
    }
}